### PR TITLE
libelf: Add patch to prevent -Wundef warnings.

### PR DESCRIPTION
**Pull request by ambrop72**
The Linux kernel includes libelf headers with -Wundef and -Werror enabled, resulting in an error when the headers are found through -I as opposed to -isystem.

Fixes https://github.com/NixOS/nixpkgs/issues/59929

###### Things done

Tested doing the things in the linked issue and confirmed the issue is gone with this patch. Note that this included a rebuild of various things including gcc.

Note that the libelf upstream site is dead.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by ambrop72** [comment]
Note that the new code is not 100% equivalent (it has different meaning when the macro is defined to 0), but these macros are either defined to 1 or undefined by a generated header, so it does not matter.

**Comment by FRidh** [comment]
Waiting for reviews before merging into staging.

**Review by matthewbauer — APPROVED** [review]

**Pull request merged**
